### ISSUE TITLE
Hex upgrades

### ISSUE THREAD

**Issue by waymost** [18xx-maker/schemas#7]
Currently, tile upgrades are specified in the `upgrades` object at the root level of game definitions as such:

```
"upgrades": {
    "3": ["141", "142", "143"],
    "4": ["141", "142"],
    "5": ["14", "15"],
    // ...
}
```

However, this neglects to specify what tiles can be laid on map hexes initially. There are potential naming conflicts between map coordinates and tile ids as some tiles in certain games are identified by a single letter followed by a number.

I propose to add an `upgrades` array of objects under `map` that would look like so:

```
"map": {
    "upgrades": [
        { "hexes": ["H16"], "upgrades": ["5", "6", "57"], "exclusions": ["14", "63", "611"] }
    ]
}
```

The `hexes` property specifies which hexes the object is defining tile upgrades for. The `upgrades` property specifies what tiles can be laid directly on to the specified hexes. The `exclusions` property specifies what tiles can't be laid on this hex even if they would be typically allowed by tile upgrades. The example above is for Savannah in 1832. Since it has water on two sides, there's no legal way to lay a 14, 63, or 611 on it, even though looking at tile upgrades on a city hex  would normally allow them.

Also, having a separate `upgrades` array under `map` prevents issues of having to split up hex definitions into multiple objects just to support different tile upgrades (e.g. border hexes).

**Comment by waymost** [comment]
We could also potentially get rid of the separate `upgrades` object for tile. Since `tiles` and `upgrades` use the same keys and `tiles` supports object values, `upgrades` can be rolled in like so.

```
"tiles": [
    "5": { "quantity": 2, "upgrades": ["14", "15"] }
]
```

This preserves backwards compatibility with how `tiles` is supported by 18xx Maker right now as far as I know, gets rid of the extra object at the root level, and remains easy to input. That said, this isn't needed for anything and the existing schema for `upgrades` can be left as is from a functionality standpoint.

**Comment by kelsin** [comment]
I like leaving it off of tiles so that tiles can still be super simple when you ONLY want to reference a known tile. Otherwise we'll be making tiles always have to be an object even when you don't need it (otherwise) to be.

**Comment by kelsin** [comment]
Matt if you add upgrades to a game file I can fix the schema after etc.

**Comment by waymost** [comment]
Yeah, after I gave it some more thought, I suspected that leaving them off tiles made more sense for the reason you said. Don't want to force adding game-specific properties to what otherwise would be a global object.

**Comment by waymost** [comment]
Sure, I can add hex upgrades to 1832 and reference the PR here.
